plugin/db/mysql/binlog.go

```diff
@@ -0,0 +1,265 @@
+		}
+	} else {
+		// The binlog event type is none of DELETE/UPDATE/INSERT. Skip for now.
+		rowEvent = nil
+	}
+
+	return rowEvent, nil
+}
+
+func parseBinlogEventQuery(header string) (*binlogEvent, error) {
+	matches := regexpThreadID.FindStringSubmatch(header)
+	if len(matches) != 2 {
+		return nil, errors.Errorf("failed to parse thread ID from the QUERY event header: %q", header)
+	}
+	return &binlogEvent{
+		Type:     QueryEvent,
+		ThreadID: matches[1],
+	}, nil
+}
+
+func parseBinlogEventDML(eventType binlogEventType, body []string) (*binlogEvent, error) {
+	if len(body) < eventType.MinBlockLen() {
+		return nil, errors.Errorf("invalid %s event body, must be at least %d lines", eventType.String(), eventType.MinBlockLen())
+	}
+	groups, err := splitBinlogEventBody(body, eventType.String())
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to split %s event body", eventType.String())
+	}
+
+	rowEvent := &binlogEvent{
+		Type: eventType,
+	}
+	re := eventType.Regexp()
+	if re == nil {
+		// Should not reach here.
+		return nil, errors.Errorf("invalid DML binlog event type %s", eventType.String())
+	}
+	for i, block := range groups {
+		if len(block) < eventType.MinBlockLen() {
+			return nil, errors.Errorf("binlog event payload must be at least %d lines, bot got %q", eventType.MinBlockLen(), strings.Join(block, "\n"))
+		}
+		if i == 0 {
+			matches := re.FindStringSubmatch(block[0])
+			if len(matches) != 3 {
+				return nil, errors.Wrapf(err, "failed to parse database and table names from the binlog event %q with regexp %s", block[0], re.String())
+			}
+			rowEvent.Database = matches[1]
+			rowEvent.Table = matches[2]
+		}
+		beforeValue, afterValue, err := eventType.ParseDMLPayload(block)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to parse the DML binlog event payload")
+		}
+		if beforeValue != nil {
+			rowEvent.DataBefore = append(rowEvent.DataBefore, beforeValue)
+		}
+		if afterValue != nil {
+			rowEvent.DataAfter = append(rowEvent.DataAfter, afterValue)
+		}
+	}
+
+	return rowEvent, nil
+}
+
+func parseBinlogEventPayloadBlock(lines []string, header string) ([]string, error) {
+	if lines[0] != fmt.Sprintf("### %s", header) {
+		return nil, errors.Errorf("failed to parse event payload head line, expecting \"### %s\", but got %q", header, lines[0])
+	}
+	var values []string
+	for i, line := range lines[1:] {
+		prefix := fmt.Sprintf("###   @%d=", i+1)
+		if !strings.HasPrefix(line, prefix) {
+			return nil, errors.Errorf("invalid binlog event payload line %q, expecting prefix %q", line, prefix)
+		}
+		values = append(values, strings.TrimPrefix(line, prefix))
+	}
+	return values, nil
+}
+
+func splitBinlogEventBody(lines []string, prefix string) ([][]string, error) {
+	var groups [][]string
+	var group []string
+	for _, line := range lines {
+		if !strings.HasPrefix(line, "### ") {
+			return nil, errors.Errorf("invalid event payload line %q, must starts with \"### \"", line)
```

Comment:
must start

Reply:
The eventType is parsed later. Here we are only grouping the string.